### DISCUSSION TITLE
Require python code for clkout functions on gpio pin for 1MhZ or above on RPI pico

### DISCUSSION THREAD

**Discussion by gurusar** [micropython/micropython#9269] [Core Development]
Need to generate a continuous square wave using internal clock frequency  on the GPIO pin(not the PWM) as clkout function on RPI pico, which  is avalaible in rp2040 as per datasheet. What will be the Thonny code for the same as i have surfed the internet but did not get any such code or help. Please if any one can share code or any documents regarding this topic.

**Comment by robert-hh** [comment]
The easiest way to get that is using the PIO. The most basic example for PIO is a square wave generator. The maximum frequency you can achieve is Fcpu / 2, e.g. 62.5 MHz at the standard CPU frequency of 125 MHz.  The code for creating 1 MHz at GPIO2 could look like:
```
@rp2.asm_pio(set_init=rp2.PIO.OUT_LOW)
def square():
    set(pins, 1)
    set(pins, 0)

# Create and start a StateMachine with square, outputting on Pin(2)
sm = rp2.StateMachine(0, square, freq=2_000_000, set_base=Pin(2))
sm.active(1)
```
The freq at the pin is half the frequency set for the PIO. So the second-to-last line sets Pin and frequency.

**Comment by dpgeorge** [comment]
If you want to use the `clk_gpout` functionality of the rp2040 clock tree, you'll need to do writes to the peripheral registers using `machine.mem32`.  You'll need to implement the functionality of this C function: https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_clocks/clocks.c#L317-L337

**Reply by dpgeorge** [reply]
Try something like this (untested!):
```py
import machine

CLOCKS_BASE = 0x40008000
CLOCKS_CTRL = 0
CLOCKS_DIV = 4
CLOCKS_CLK_GPOUT0_CTRL_AUXSRC_LSB = 5
CLOCKS_CLK_GPOUT0_CTRL_ENABLE_BITS = 0x00000800
CLOCKS_CLK_GPOUT0_DIV_INT_LSB = 8
GPIO_FUNC_GPCK = 8

def clock_gpio_init(pin, src, div):
    gpio_id = str(pin).split(", ")[0][4:]
    gpclk = (0, None, 1, 2, 3)[gpio_id - 21]
    machine.mem32[CLOCKS_BASE + gpclk * 12 + CLOCKS_CTRL] = src << CLOCKS_CLK_GPOUT0_CTRL_AUXSRC_LSB | CLOCKS_CLK_GPOUT0_CTRL_ENABLE_BITS
    machine.mem32[CLOCKS_BASE + gpclk * 12 + CLOCKS_DIV] = div << CLOCKS_CLK_GPOUT0_DIV_INT_LSB
    pin.init(pin.ALT, alt=GPIO_FUNC_GPCK)
    
    
# src is clk_sys, div is 1
clock_gpio_init(machine.Pin(21), 0x6, 0x100)
```

**Reply by robert-hh** [reply]
That's a neat code. Just a small change in line 12:
```
import machine

CLOCKS_BASE = 0x40008000
CLOCKS_CTRL = 0
CLOCKS_DIV = 4
CLOCKS_CLK_GPOUT0_CTRL_AUXSRC_LSB = 5
CLOCKS_CLK_GPOUT0_CTRL_ENABLE_BITS = 0x00000800
CLOCKS_CLK_GPOUT0_DIV_INT_LSB = 8
GPIO_FUNC_GPCK = 8

def clock_gpio_init(pin, src, div):
    gpio_id = int(str(pin).split(", ")[0][4:])
    gpclk = (0, None, 1, 2, 3)[gpio_id - 21]
    machine.mem32[CLOCKS_BASE + gpclk * 12 + CLOCKS_CTRL] = src << CLOCKS_CLK_GPOUT0_CTRL_AUXSRC_LSB | CLOCKS_CLK_GPOUT0_CTRL_ENABLE_BITS
    machine.mem32[CLOCKS_BASE + gpclk * 12 + CLOCKS_DIV] = div << CLOCKS_CLK_GPOUT0_DIV_INT_LSB
    pin.init(pin.ALT, alt=GPIO_FUNC_GPCK)
    
    
# src is clk_sys, div is 1
clock_gpio_init(machine.Pin(21), 0x6, 0x100)
```
It outputs 488.2822 kHz at Pin 21, or Fcpu / 256.

**Reply by robert-hh** [reply]
With `clock_gpio_init(machine.Pin(21), 0x6, 0x100)` changed to `clock_gpio_init(machine.Pin(21), 0x6, 0x80)`
and machine.freq() set to `machine.freq(128000000)` it's 1 MHz.
edit:
or of course;
`clockout.clock_gpio_init(machine.Pin(21), 0x6, 125)` at the default clock frequency.

**Reply by robert-hh** [reply]
Just a comparison Pio output vs. clockout. The yellow trace is from the PIO, the green one from the clockout. Almost no phase jitter.

![pio_vs_clockout](https://user-images.githubusercontent.com/12476868/189518296-41d8f9f1-fc01-4f05-b36c-c44eb161b1eb.jpg)

**Reply by rkompass** [reply]
`clock_gpio_init(machine.Pin('LED'), 10, 12_000_000)` now is a new way to let the LED blink at 1 Hz.

**Reply by gurusar** [reply]
Thank you all for your valuable replies.
I have copy and pasted the code in Thonny and tried to run the the code ,But didin,t get any output wave at pin 21 of RPI PICO W board. According to the function defined system clock is divided by 4 and system clock is 125Mhz as defined inregisters.
but acc. to void clock_gpio_init(uint gpio, uint src, uint div) definitely the frequency should be 488khz. 
So, Please if anyone can add comments after each line, as it's not clear even in datasheet where .c file is given.
Even i have seen the .c code given at github and int the datasheet of rp2040 , python code written is almost same. 
Please see if am doing some thing wrong or missing something.

**Reply by rkompass** [reply]
I will try to explain.
But I don't understand your problem. The above corrected code runs from Thonny here (copy and paste and a save on local disk, of course needed). With `clock_gpio_init(machine.Pin(21), 0x6, 125)` it gives 1 MHz at pin 21. The system clock 125 MHz is divided by 125.

I go through the code lines in the function:
`gpio_id = int(str(pin).split(", ")[0][4:])` 
extracts the pin number from the machine.Pin object. Assume you have defined `pin = machine.Pin(21)` then `str(pin)` gives you `'Pin(21, mode=ALT, pull=PULL_DOWN, alt=31)'`.
`str(pin).split(",")` returns the list of substrings `['Pin(21', ' mode=ALT', ' pull=PULL_DOWN', ' alt=31)']` and from the first element `'Pin(21'` the substring from character 4 to the end is extracted by indexing `[0][4:]`. Finally from this substring `'21'` the integer 21 is computed via `int()`.

`gpclk = (0, None, 1, 2, 3)[gpio_id - 21]`
does the following mapping  21 -> 0, 22-> None, 23->1, 24->2, 25 -> 3. So instead of pin number 21 we have now gpout clock channel number 0.

`machine.mem32[CLOCKS_BASE + gpclk * 12 + CLOCKS_CTRL] = src << CLOCKS_CLK_GPOUT0_CTRL_AUXSRC_LSB | CLOCKS_CLK_GPOUT0_CTRL_ENABLE_BITS`
reduces to
`machine.mem32[0x40008000 + gpclk * 12] = 0x800 | src << 5`
after inserting the constants. With gpclk == 0 in case of pin 21 we have a memory write at location 0x40008000 which is the 
 CLK_GPOUT0_CTRL register (see page 222 of the RPI2040 datasheet pdf). The enable bit 11 (0x800 is just a 1 at position 11) is set and the src variable is written at bits 8..5. In the pdf you find values for other source values.

The next instruction translates in similar manner to a memory write at location 0x40008004 which is the CLK_GPOUT0_DIV register.
The value `div <<8`  is written into it. So the integer component  of the divider variable is placed at bits 31..8.
You may change this to `round(div * 256)` to preserve the ability to have fractional dividers.

The last line of the function sets the GPIO Pin function to 8, which is Clock GPIN or GPOUT at certain pins (20-25) only.
See page 262 of the datasheet for this. The pin thus is multiplexed to output the now enabled and configured internal clock.

In experimenting yesterday I found that with switching the clock in several runs of `clock_gpio_init()` the clock may be missing apparently.
If you have a big divider at a fast clock and then switch to a smaller divider at a slower clock it seems, the new divider is only loaded after the now slower clock has passed through all remaining cycles of the old divider.
The other clock sources I found interesting are:
```
6 → clk_sys  (125 MHz)
7 → clk_usb   (48 MHz)
8 → clk_adc   (48 MHz)
9 → clk_rtc (46875 Hz)
10 → clk_ref  (12 MHz)
```



I have an ordinary RPI Pico here and a late uPy version running. As far as I can see, pin 21 should behave the same on the Pico W.
Do you have a possibility to try out on an ordinary Pico?
Then `clock_gpio_init(machine.Pin('LED'), 9, 93750)` should let the LED blink at 0.5 Hz. On the Pico W this should not work, as the LED is connected differently now. But pin 21 should work identically.

**Reply by gurusar** [reply]
Thank you for reply again.
Actually ,I am using RPI PI PICO readymade board available in market. They have not provided the pin 21 at outer peripheral pins . Secondly the board is mounted with a smd crystal on pins 21 and 20 . However i have tested with a DSO on the pin 21 of RP2040 but no output . This might happen due to crystal also mounted on the board on the same pin 21. I have to check the code with my own mounted chip with my own pcb design i think. Do you have any suggestions on this or  am i going in wrong direction?.

**Reply by rkompass** [reply]
![Pi-Pico-Pinout](https://user-images.githubusercontent.com/90282516/189942510-0b98831e-2ae0-4ac2-ba4e-e2692ddb800f.jpeg)
This is the board I use. It is the standard one. I'm not sure I understood you fully. By "readymade board" do you refer to this too? If not I suggest you buy such one, it's only 5$. Once you have success measuring the signal you should start with deviations like own pcb, otherwise you do not know where to locate any problem that might appear.

Perhaps there was a misunderstanding about pins? `machine.Pin(21)` refers to GP21 on the pinout graphics. At least with this board there is no crystal between GP20 and GP21.

In the docs there is a way to construct a relaxation oscillator described. This is done by a resistor connecting GP20 and GP21 and a small capacitor from GP20 (input) to GND. Both these pins have to be set to alternate function mode 8 and the output (GP21) also has to be set to inverted mode in the appropriate GPIO configuration register. I suppose you can use the above `clock_gpio_init()` init function for setting divider at 1 and the source to 1 (clksrc_gpin0). Then you have essentially an inverter with a RC-delay to generate the frequency.
It would be interesting to try this out. If you are interested in this, I could try that.

A short google [showed](https://electronics.stackexchange.com/questions/218142/using-cmos-schmitt-trigger-inverters-in-quartz-crystal-oscillator-circuit) that instead of the RC-part one might use a quartz crystal. Because in the above configuration GP20 and GP21 behave like input and output of a logic inverter. The oscillation would become much more stable than with RC.
Perhaps this is the wiring of your board? Then this crystal oscillator should perhaps be configured like a relaxation oscillator.
And without such configuration there would be no oscillation.

**Reply by gurusar** [reply]
Thank you for explanation, That was a great help .Yes i have the same board. I have tested gp21 as in schematic and got the wave on gpin 21. I was checking Xout pin on rp2040 ,which is pin 21 of microcontroller. This pin is ideally used to generate clock out if using clock out in microcontroller as per datasheet. Well i have got the output.  Now i am going to explore (write big tables in internal eeprom and read the frequency on  particular pin and display on LCD). Hope it go smooth. Thank you once again.
